libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp

```diff
@@ -0,0 +1,312 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Yul interpreter module that evaluates EVM instructions.
+ */
+
+#include <libyul/tools/interpreter/PureEVMInstructionInterpreter.h>
+
+#include <libyul/tools/interpreter/PureInterpreter.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AST.h>
+#include <libyul/Utilities.h>
+
+#include <libevmasm/Instruction.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/Keccak256.h>
+#include <libsolutil/Numeric.h>
+#include <libsolutil/picosha2.h>
+
+#include <boost/algorithm/string.hpp>
+
+using namespace solidity;
+using namespace solidity::evmasm;
+using namespace solidity::yul;
+using namespace solidity::yul::tools::interpreter;
+
+using solidity::util::h160;
+using solidity::util::h256;
+
+using u512 = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<512, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>;
```

Comment:
Why not place this next to the other types like this we already have, like `u256`?

Reply:
I did not notice that type was there. I was too focused on modifying the original interpreter :pray: .

I will move it.

Reply:
I have move `u512` to `libsolutil/numeric`. However, I have found a minor thing with the type definition: the `MaxBits` is `256`, while `MinBits` is 512. This is fixed in the commit https://github.com/ethereum/solidity/pull/15464/commits/5351aa8905dc12b4b1a603236adb2b78a59bf8ac .

As I was copy and pasting the implementation before adding my change, the original interpreter in the `test/` dir has the same _issue_.

Currently the _issue_ does not lead to any error. I checked boost implementation, and the precision is obtained by getting the max of `MaxBits` and `MinBits` (if the allocator is `void`). I think maybe we can just use the predefined types from [boost cpp_int](https://www.boost.org/doc/libs/1_84_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html) instead of redefined like this.

Reply:
> I have move `u512` to `libsolutil/numeric`. However, I have found a minor thing with the type definition: the `MaxBits` is `256`, while `MinBits` is 512. This is fixed in the commit https://github.com/ethereum/solidity/pull/15464/commits/5351aa8905dc12b4b1a603236adb2b78a59bf8ac .

Nice. Can you submit it as a separate PR (along with moving `u512` to `libsolutil`)? We can merge it as a bugfix.

> I think maybe we can just use the predefined types from [boost cpp_int](https://www.boost.org/doc/libs/1_84_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html) instead of redefined like this.

Sure, why not. I'm not sure why this wasn't done in the first place. Maybe these typedefs did not exist back when we defined these types?

You can submit this as a PR too.

Reply:
I have made the PR here #15510